src/webgpu/shader/execution/builtin/fract.spec.ts

```diff
@@ -0,0 +1,63 @@
+export const description = `
+Execution Tests for the 'fract' builtin function
+`;
+
+import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { GPUTest } from '../../../gpu_test.js';
+import { f32, f32Bits, TypeF32 } from '../../../util/conversion.js';
+
+import { Config, correctlyRoundedThreshold, kBit, run } from './builtin.js';
+
+export const g = makeTestGroup(GPUTest);
+
+g.test('float_builtin_functions,fract')
+  .uniqueId('58222ecf6f963798')
+  .specURL('https://www.w3.org/TR/2021/WD-WGSL-20210929/#float-builtin-functions')
+  .desc(
+    `
+fract:
+T is f32 or vecN<f32> fract(e: T ) -> T Returns the fractional bits of e (e.g. e - floor(e)). Component-wise when T is a vector. (GLSLstd450Fract)
+`
+  )
+  .params(u =>
+    u
+      .combine('storageClass', ['uniform', 'storage_r', 'storage_rw'] as const)
+      .combine('vectorize', [undefined, 2, 3, 4] as const)
+  )
+  .fn(async t => {
+    const cfg: Config = t.params;
+    cfg.cmpFloats = correctlyRoundedThreshold();
+    run(t, 'fract', [TypeF32], TypeF32, cfg, [
+      // Positive numbers
+      { input: f32Bits(0x3dcccccd), expected: f32Bits(0x3dcccccd) }, // ~0.1 -> ~0.1
+      { input: f32(0.5), expected: f32(0.5) }, // 0.5 -> 0.5
+      { input: f32Bits(0x3f666666), expected: f32Bits(0x3f666666) }, // ~0.9 -> ~0.9
+      { input: f32Bits(0x3f800000), expected: f32(0) }, // 1 -> 0
+      { input: f32Bits(0x40000000), expected: f32(0) }, // 2 -> 0
+      { input: f32Bits(0x3f8e147b), expected: f32Bits(0x3de147b0) }, // ~1.11 -> ~0.11
+      { input: f32Bits(0x41200069), expected: f32Bits(0x38d20000) }, // ~10.0001 -> ~0.0001
+
+      // Negative numbers
+      { input: f32Bits(0xbdcccccd), expected: f32Bits(0x3f666666) }, // ~-0.1 -> ~0.9
+      { input: f32(-0.5), expected: f32(0.5) }, // -0.5 -> 0.5
+      { input: f32Bits(0xbf666666), expected: f32Bits(0x3dccccd0) }, // ~-0.9 -> ~0.1
+      { input: f32Bits(0xbf800000), expected: f32(0) }, // -1 -> 0
+      { input: f32Bits(0xc0000000), expected: f32(0) }, // -2 -> 0
+      { input: f32Bits(0xbf8e147b), expected: f32Bits(0x3f63d70a) }, // ~-1.11 -> ~0.89
+      { input: f32Bits(0xc1200419), expected: f32Bits(0x3f7fbe70) }, // ~-10.0001 -> ~0.999
+
+      // Min and Max f32
+      { input: f32Bits(kBit.f32.positive.min), expected: f32Bits(kBit.f32.positive.min) },
+      { input: f32Bits(kBit.f32.positive.max), expected: f32(0) },
+      { input: f32Bits(kBit.f32.negative.max), expected: f32Bits(0x3f7fffff) },
+      { input: f32Bits(kBit.f32.negative.min), expected: f32(0) },
+
+      // Subnormal f32
+      // prettier-ignore
+      { input: f32Bits(kBit.f32.subnormal.positive.max), expected: f32Bits(kBit.f32.subnormal.positive.max) },
```

Comment:
https://www.w3.org/TR/WGSL/#floating-point-evaluation states that:

> Implementations may flush denormalized value on the input and/or output of any operation listed in [§ 12.5.1 Floating Point Accuracy](https://www.w3.org/TR/WGSL/#floating-point-accuracy).
> Other operations are required to preserve denormalized numbers.

I assume "any operation listed in [§ 12.5.1 Floating Point Accuracy](https://www.w3.org/TR/WGSL/#floating-point-accuracy)." includes the contents of the table _"Accuracy of built-in functions"_, and so `floor` is allowed to flush denormalized values to zero. @dneto0 or @alan-baker to confirm†.

I don't see any handling of denormals in `correctlyRounded()`, so I assume that this isn't handled by the tests. You could change `correctlyRounded()`, or change `expected` to `[f32Bits(kBit.f32.subnormal.positive.max), f32(0)]`

---
† Also: would it be clearer to change:
> Implementations may flush denormalized value...

to

> Implementations may **flush-to-zero** denormalized value...

As not everyone reading may know what the value is flushed to?

Reply:
Correct, floor can flush denormalized values.

Reply:
I should probably update correctlyRounded to flush denormals, so that subsequent tests don't have to be away of this subtleness.

Related to that, currently nextAfter, which is used be correctlyRounded(0) will return the subnormal nearest to 0 that is in the appropriate direction. Is it correct to assume that nextAfter should also flush subnormals? thus this should be changed to the nearest normal to 0, otherwise we will get into a potential situation of getting trapped at 0.